File: app/src/main/java/com/jojokorok/nourishrx/medications/MedicationScreens.java
package com.jojokorok.nourishrx.medications;

import android.app.Activity;
import android.graphics.Typeface;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.LinearLayout;

import com.jojokorok.nourishrx.data.Medication;
import com.jojokorok.nourishrx.data.MedicationStore;
import com.jojokorok.nourishrx.ui.NourishColors;
import com.jojokorok.nourishrx.ui.NourishUi;

import java.util.List;

public class MedicationScreens {
    public interface Callbacks {
        long currentProfileId();

        String selectedProfileName();

        View sectionTitle(String title, String subtitle);

        void emptyState(String message, String action, View.OnClickListener listener);

        void showMedicationDialog(Medication medication);

        void toggleMedication(Medication medication);

        void confirmDelete(Medication medication);

        void adjustInventory(Medication medication, int delta);

        void showInventoryDialog(Medication medication);
    }

    private final Activity activity;
    private final MedicationStore store;
    private final NourishUi ui;
    private final Callbacks callbacks;

    public MedicationScreens(
            Activity activity,
            MedicationStore store,
            NourishUi ui,
            Callbacks callbacks
    ) {
        this.activity = activity;
        this.store = store;
        this.ui = ui;
        this.callbacks = callbacks;
    }

    public void renderMedications(LinearLayout content) {
        List<Medication> medications = store.getAllMedications(callbacks.currentProfileId());
        String profileName = callbacks.selectedProfileName();
        content.addView(callbacks.sectionTitle("Medications", profileName + " has " + medications.size() + " saved"));

        if (medications.isEmpty()) {
            callbacks.emptyState("Add names, doses, instructions, and reminders for " + profileName + ".", "Add medication", view -> callbacks.showMedicationDialog(null));
            return;
        }

        for (Medication medication : medications) {
            content.addView(medicationCard(medication));
        }
    }

    public void renderInventory(LinearLayout content) {
        List<Medication> medications = store.getAllMedications(callbacks.currentProfileId());
        long lowCount = medications.stream().filter(Medication::isLowStock).count();
        String profileName = callbacks.selectedProfileName();
        content.addView(callbacks.sectionTitle("Stock", profileName + " has " + lowCount + " low stock"));

        if (medications.isEmpty()) {
            callbacks.emptyState("Inventory for " + profileName + " appears here after adding meds.", "Add medication", view -> callbacks.showMedicationDialog(null));
            return;
        }

        for (Medication medication : medications) {
            content.addView(inventoryCard(medication));
        }
    }

    private View medicationCard(Medication medication) {
        LinearLayout card = ui.card();
        LinearLayout top = new LinearLayout(activity);
        top.setOrientation(LinearLayout.HORIZONTAL);
        top.setGravity(Gravity.CENTER_VERTICAL);

        LinearLayout details = new LinearLayout(activity);
        details.setOrientation(LinearLayout.VERTICAL);
        details.addView(ui.text(medication.name, 19, NourishColors.INK, Typeface.BOLD));
        details.addView(ui.text(medication.dosage, 14, NourishColors.MUTED, Typeface.NORMAL));
        details.addView(ui.text(medication.doseCountLabel() + " at " + medication.scheduleSummary(), 14, NourishColors.MUTED, Typeface.NORMAL));
        details.addView(ui.text(medication.repeatReminderLabel(), 13, NourishColors.MUTED, Typeface.NORMAL));
        if (!medication.instructions.isEmpty()) {
            details.addView(ui.text(medication.instructions, 14, NourishColors.MUTED, Typeface.NORMAL));
        }
        top.addView(details, new LinearLayout.LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT, 1));
        top.addView(ui.statusBadge(medication.active ? "Active" : "Paused"));
        card.addView(top);

        LinearLayout actions = actionRow();
        Button edit = button("Edit", NourishColors.BLUE, NourishColors.BLUE_SOFT);
        edit.setOnClickListener(view -> callbacks.showMedicationDialog(medication));
        actions.addView(edit, weightedActionParams());

        Button toggle = button(medication.active ? "Pause" : "Resume", NourishColors.GREEN, NourishColors.GREEN_SOFT);
        toggle.setOnClickListener(view -> callbacks.toggleMedication(medication));
        actions.addView(toggle, weightedActionParams());

        Button delete = button("Delete", NourishColors.CORAL, NourishColors.CORAL_SOFT);
        delete.setOnClickListener(view -> callbacks.confirmDelete(medication));
        actions.addView(delete, weightedActionParams());
        card.addView(actions);
        return card;
    }

    private View inventoryCard(Medication medication) {
        LinearLayout card = ui.card();
        LinearLayout top = new LinearLayout(activity);
        top.setOrientation(LinearLayout.HORIZONTAL);
        top.setGravity(Gravity.CENTER_VERTICAL);

        LinearLayout details = new LinearLayout(activity);
        details.setOrientation(LinearLayout.VERTICAL);
        details.addView(ui.text(medication.name, 19, NourishColors.INK, Typeface.BOLD));
        details.addView(ui.text(
                medication.quantity + " remaining",
                14,
                medication.isLowStock() ? NourishColors.CORAL : NourishColors.MUTED,
                Typeface.BOLD
        ));
        details.addView(ui.text("Refill threshold: " + medication.refillThreshold, 13, NourishColors.MUTED, Typeface.NORMAL));
        top.addView(details, new LinearLayout.LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT, 1));
        top.addView(ui.statusBadge(medication.isLowStock() ? "Refill" : "OK"));
        card.addView(top);

        LinearLayout actions = actionRow();
        Button minus = button("-1", NourishColors.CORAL, NourishColors.CORAL_SOFT);
        minus.setOnClickListener(view -> callbacks.adjustInventory(medication, -1));
        actions.addView(minus, weightedActionParams());

        Button plus = button("+10", NourishColors.GREEN, NourishColors.GREEN_SOFT);
        plus.setOnClickListener(view -> callbacks.adjustInventory(medication, 10));
        actions.addView(plus, weightedActionParams());

        Button set = button("Set", NourishColors.BLUE, NourishColors.BLUE_SOFT);
        set.setOnClickListener(view -> callbacks.showInventoryDialog(medication));
        actions.addView(set, weightedActionParams());
        card.addView(actions);
        return card;
    }

    private LinearLayout actionRow() {
        LinearLayout actions = new LinearLayout(activity);
        actions.setOrientation(LinearLayout.HORIZONTAL);
        actions.setPadding(0, ui.dp(12), 0, 0);
        return actions;
    }

    private LinearLayout.LayoutParams weightedActionParams() {
        LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(0, ui.dp(44), 1);
        params.rightMargin = ui.dp(8);
        return params;
    }

    private Button button(String label, int textColor, int backgroundColor) {
        return ui.button(label, textColor, backgroundColor);
    }
}

File: app/src/main/java/com/jojokorok/nourishrx/medications/MedicationScreens.kt
package com.jojokorok.nourishrx.medications

import android.app.Activity
import android.graphics.Color
import android.graphics.Typeface
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.LinearLayout
import com.jojokorok.nourishrx.data.Medication
import com.jojokorok.nourishrx.data.MedicationStore
import com.jojokorok.nourishrx.ui.NourishColors
import com.jojokorok.nourishrx.ui.NourishShapes
import com.jojokorok.nourishrx.ui.NourishSpacing
import com.jojokorok.nourishrx.ui.NourishTypography
import com.jojokorok.nourishrx.ui.NourishUi

class MedicationScreens(
    private val activity: Activity,
    private val store: MedicationStore,
    private val ui: NourishUi,
    private val callbacks: Callbacks
) {
    interface Callbacks {
        fun currentProfileId(): Long
        fun showMedicationDialog(medication: Medication?)
        fun toggleMedication(medication: Medication)
        fun confirmDelete(medication: Medication)
        fun adjustInventory(medication: Medication, delta: Int)
        fun showInventoryDialog(medication: Medication)
    }

    fun renderMedications(content: LinearLayout) {
        val medications = store.getAllMedications(callbacks.currentProfileId())
        val activeCount = medications.count { it.active }
        val pausedCount = medications.size - activeCount
        val summary = when {
            medications.isEmpty() -> "No medications saved"
            pausedCount == 0 -> plural(activeCount, "active medication", "active medications")
            else -> "$activeCount active, $pausedCount paused"
        }
        content.addView(sectionHeader("Medication library", summary))

        if (medications.isEmpty()) {
            content.addView(
                emptyState(
                    "Your medication library is empty",
                    "Save medication details, dose times, and reminder settings here."
                )
            )
            return
        }

        medications.forEach { content.addView(medicationCard(it)) }
    }

    fun renderInventory(content: LinearLayout) {
        val medications = store.getAllMedications(callbacks.currentProfileId())
        val lowCount = medications.count { it.isLowStock() }
        val summary = when {
            medications.isEmpty() -> "No stock to track"
            lowCount == 0 -> "All medication stock is above its refill level"
            else -> plural(lowCount, "medication needs a refill", "medications need a refill")
        }
        content.addView(sectionHeader("Medication stock", summary))

        if (medications.isEmpty()) {
            content.addView(
                emptyState(
                    "No inventory yet",
                    "Add a medication to start tracking quantities and refill levels."
                )
            )
            return
        }

        medications.forEach { content.addView(inventoryCard(it)) }
    }

    private fun sectionHeader(title: String, subtitle: String): LinearLayout =
        LinearLayout(activity).apply {
            orientation = LinearLayout.VERTICAL
            setPadding(0, ui.dp(NourishSpacing.XS), 0, ui.dp(NourishSpacing.XXS))
            addView(ui.displayText(title, NourishTypography.TITLE, NourishColors.INK))
            addView(
                ui.text(
                    subtitle,
                    NourishTypography.LABEL,
                    NourishColors.MUTED,
                    Typeface.NORMAL
                )
            )
        }

    private fun medicationCard(medication: Medication): LinearLayout = ui.card().apply {
        elevation = ui.dp(NourishShapes.ELEVATION_FLAT).toFloat()
        addView(cardHeader(medication, if (medication.active) "Active" else "Paused"))
        addView(divider(), dividerParams())
        addView(
            metadataRow(
                "Schedule",
                "${medication.doseCountLabel()} at ${medication.scheduleSummary()}"
            ),
            matchWrapParams(topMargin = NourishSpacing.SM)
        )
        addView(
            metadataRow("Alerts", medication.repeatReminderLabel()),
            matchWrapParams(topMargin = NourishSpacing.XS)
        )

        if (medication.instructions.isNotEmpty()) {
            addView(
                metadataRow("Instructions", medication.instructions),
                matchWrapParams(topMargin = NourishSpacing.XS)
            )
        }

        addView(divider(), dividerParams())
        addView(medicationActions(medication))
    }

    private fun inventoryCard(medication: Medication): LinearLayout = ui.card().apply {
        elevation = ui.dp(NourishShapes.ELEVATION_FLAT).toFloat()
        val status = when {
            !medication.active -> "Paused"
            medication.isLowStock() -> "Refill"
            else -> "Active"
        }
        addView(cardHeader(medication, status))
        addView(divider(), dividerParams())
        addView(stockSummary(medication), matchWrapParams(topMargin = NourishSpacing.SM))
        addView(divider(), dividerParams())
        addView(inventoryActions(medication))
    }

    private fun cardHeader(medication: Medication, status: String): LinearLayout =
        LinearLayout(activity).apply {
            orientation = LinearLayout.HORIZONTAL
            gravity = Gravity.TOP

            val details = LinearLayout(activity).apply {
                orientation = LinearLayout.VERTICAL
                addView(
                    ui.displayText(
                        medication.name,
                        NourishTypography.BODY_LARGE,
                        NourishColors.INK
                    )
                )
                if (medication.dosage.isNotEmpty()) {
                    addView(
                        ui.text(
                            medication.dosage,
                            NourishTypography.LABEL,
                            NourishColors.INK_SECONDARY,
                            Typeface.NORMAL
                        )
                    )
                }
            }
            addView(details, LinearLayout.LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT, 1f))
            addView(
                ui.statusBadge(status),
                LinearLayout.LayoutParams(
                    ViewGroup.LayoutParams.WRAP_CONTENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT
                ).apply {
                    leftMargin = ui.dp(NourishSpacing.SM)
                }
            )
        }

    private fun metadataRow(label: String, value: String): LinearLayout =
        LinearLayout(activity).apply {
            orientation = LinearLayout.HORIZONTAL
            gravity = Gravity.TOP
            addView(
                ui.text(
                    label,
                    NourishTypography.CAPTION,
                    NourishColors.MUTED,
                    Typeface.BOLD
                ),
                LinearLayout.LayoutParams(ui.dp(88), ViewGroup.LayoutParams.WRAP_CONTENT).apply {
                    rightMargin = ui.dp(NourishSpacing.XS)
                }
            )
            addView(
                ui.text(
                    value,
                    NourishTypography.LABEL,
                    NourishColors.INK_SECONDARY,
                    Typeface.NORMAL
                ),
                LinearLayout.LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT, 1f)
            )
        }

    private fun stockSummary(medication: Medication): LinearLayout =
        LinearLayout(activity).apply {
            orientation = LinearLayout.HORIZONTAL
            gravity = Gravity.BOTTOM

            val quantity = LinearLayout(activity).apply {
                orientation = LinearLayout.VERTICAL
                addView(
                    ui.displayText(
                        medication.quantity.toString(),
                        NourishTypography.SCREEN_TITLE,
                        if (medication.isLowStock()) NourishColors.CORAL else NourishColors.GREEN_DARK
                    )
                )
                addView(
                    ui.text(
                        "remaining",
                        NourishTypography.LABEL,
                        NourishColors.MUTED,
                        Typeface.NORMAL
                    )
                )
            }
            addView(quantity, LinearLayout.LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT, 1f))

            val threshold = LinearLayout(activity).apply {
                orientation = LinearLayout.VERTICAL
                gravity = Gravity.END
                addView(
                    ui.text(
                        "Refill at",
                        NourishTypography.CAPTION,
                        NourishColors.MUTED,
                        Typeface.BOLD
                    )
                )
                addView(
                    ui.displayText(
                        medication.refillThreshold.toString(),
                        NourishTypography.BODY_LARGE,
                        NourishColors.INK_SECONDARY
                    )
                )
            }
            addView(threshold, LinearLayout.LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT, 1f))
        }

    private fun medicationActions(medication: Medication): LinearLayout = actionRow().apply {
        addView(
            actionButton("Edit", NourishColors.BLUE, NourishColors.BLUE_SOFT).apply {
                setOnClickListener { callbacks.showMedicationDialog(medication) }
            },
            weightedActionParams(hasTrailingMargin = true)
        )
        addView(
            actionButton(
                if (medication.active) "Pause" else "Resume",
                if (medication.active) NourishColors.INK_SECONDARY else NourishColors.GREEN,
                Color.TRANSPARENT
            ).apply {
                setOnClickListener { callbacks.toggleMedication(medication) }
            },
            weightedActionParams(hasTrailingMargin = true)
        )
        addView(
            actionButton("Delete", NourishColors.CORAL, Color.TRANSPARENT).apply {
                setOnClickListener { callbacks.confirmDelete(medication) }
            },
            weightedActionParams(hasTrailingMargin = false)
        )
    }

    private fun inventoryActions(medication: Medication): LinearLayout = actionRow().apply {
        addView(
            actionButton("-1", NourishColors.CORAL, Color.TRANSPARENT).apply {
                setOnClickListener { callbacks.adjustInventory(medication, -1) }
            },
            compactActionParams()
        )
        addView(
            actionButton("+10", NourishColors.GREEN, Color.TRANSPARENT).apply {
                setOnClickListener { callbacks.adjustInventory(medication, 10) }
            },
            compactActionParams()
        )
        addView(
            actionButton("Update stock", NourishColors.ON_ACCENT, NourishColors.GREEN).apply {
                setOnClickListener { callbacks.showInventoryDialog(medication) }
            },
            LinearLayout.LayoutParams(0, ui.dp(44), 1f)
        )
    }

    private fun emptyState(title: String, message: String): LinearLayout = ui.card().apply {
        elevation = ui.dp(NourishShapes.ELEVATION_FLAT).toFloat()

        val intro = LinearLayout(activity).apply {
            orientation = LinearLayout.HORIZONTAL
            gravity = Gravity.CENTER_VERTICAL
        }
        intro.addView(
            ui.text("Rx", NourishTypography.BODY, NourishColors.GREEN_DARK, Typeface.BOLD).apply {
                gravity = Gravity.CENTER
                background = ui.rounded(
                    NourishColors.GREEN_SOFT,
                    Color.TRANSPARENT,
                    ui.dp(NourishShapes.RADIUS_CONTROL)
                )
            },
            LinearLayout.LayoutParams(ui.dp(42), ui.dp(42)).apply {
                rightMargin = ui.dp(NourishSpacing.SM)
            }
        )
        intro.addView(
            LinearLayout(activity).apply {
                orientation = LinearLayout.VERTICAL
                addView(ui.displayText(title, NourishTypography.BODY_LARGE, NourishColors.INK))
                addView(
                    ui.text(
                        message,
                        NourishTypography.LABEL,
                        NourishColors.MUTED,
                        Typeface.NORMAL
                    )
                )
            },
            LinearLayout.LayoutParams(0, ViewGroup.LayoutParams.WRAP_CONTENT, 1f)
        )
        addView(intro)

        addView(
            actionButton("Add medication", NourishColors.ON_ACCENT, NourishColors.GREEN).apply {
                setOnClickListener { callbacks.showMedicationDialog(null) }
            },
            LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ui.dp(44)).apply {
                topMargin = ui.dp(NourishSpacing.MD)
            }
        )
    }

    private fun actionRow(): LinearLayout = LinearLayout(activity).apply {
        orientation = LinearLayout.HORIZONTAL
        setPadding(0, ui.dp(NourishSpacing.SM), 0, 0)
    }

    private fun actionButton(label: String, textColor: Int, backgroundColor: Int): Button =
        ui.button(label, textColor, backgroundColor)

    private fun weightedActionParams(hasTrailingMargin: Boolean): LinearLayout.LayoutParams =
        LinearLayout.LayoutParams(0, ui.dp(44), 1f).apply {
            if (hasTrailingMargin) rightMargin = ui.dp(NourishSpacing.XS)
        }

    private fun compactActionParams(): LinearLayout.LayoutParams =
        LinearLayout.LayoutParams(ui.dp(64), ui.dp(44)).apply {
            rightMargin = ui.dp(NourishSpacing.XS)
        }

    private fun divider(): View = View(activity).apply {
        setBackgroundColor(NourishColors.DIVIDER)
    }

    private fun dividerParams(): LinearLayout.LayoutParams =
        LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ui.dp(1)).apply {
            topMargin = ui.dp(NourishSpacing.SM)
        }

    private fun matchWrapParams(topMargin: Int): LinearLayout.LayoutParams =
        LinearLayout.LayoutParams(
            ViewGroup.LayoutParams.MATCH_PARENT,
            ViewGroup.LayoutParams.WRAP_CONTENT
        ).apply {
            this.topMargin = ui.dp(topMargin)
        }

    private fun plural(count: Int, singular: String, plural: String): String =
        "$count ${if (count == 1) singular else plural}"
}
